src/Manifolds.jl

```diff
@@ -102,6 +106,94 @@ function hat(M::Manifold, x, vⁱ)
     return v
 end
 
+@doc doc"""
+    mean(M,x; weights=1/n*ones(n), stop_tol=10^-7, stop_iter=100 )
+
+computes the Riemannian center of mass also known as Karcher mean of the vector
+`x` of `n` points on the [`Manifold`](@ref) `M`. This function
+uses the gradient descent scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian center of mass.
+The general formula to compute the minimizer reads
+````math
+\argmin_{y\in\mathcal M} \frac{1}{2}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}^2(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemian [`distance`](@ref).
+
+Finally `stop_tol` denotes the minimal change between to iterates then to stop
+and `stop_iter` the maximal number of iterations. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in 
+> Afsari, B; Tron, R.; Vidal, R.: On the Convergence of Gradient
+> Descent for Finding the Riemannian Center of Mass,
+> SIAM Journal on Control and Optimization (2013), 51(3), pp. 2230–2260, 
+> doi: [10.1137/12086282X](https://doi.org/10.1137/12086282X),
+> arxiv: [1201.0925](https://arxiv.org/abs/1201.0925">1201.0925)
+"""
+function mean(M::Manifold, x::Vector{T};
+    weights=1/length(x)*ones(length(x)),
+    stop_atol=10^-7,
+    stop_iter=100
+    ) where {T}
+    iter = 0
+    y=x[1]
+    yOld = y
+    while ( ( iter==0 || distance(M,y,yOld) > stop_atol ) && (iter < stop_iter) )
+        iter += 1
+        yOld = y
+        y = exp(M,yOld,  1/2*sum( weights.*log.(Ref(M), Ref(yOld),x) )  )
```

Comment:
I'd suggest using mutating `exp` and `log` here. Otherwise there will be a lot of allocations. Or maybe we can have two variants in case someone has some reasons for avoiding in-place operations (like `mean(::Manifold, x::AbstractVector, use_mutation::Val{true}`)

Reply:
Also `sum(...)/2` instead of `1/2*sum(...)` would avoid promotion to `Float64`. Similarly in weight initialization.

Reply:
Is the `Ref` necessary?

Reply:
Of course we can overload a a mutating one, I will check for that. Well you can not broadcast the `M` nor the `yOld` so I put them in `Ref`s?

Reply:
> Also `sum(...)/2` instead of `1/2*sum(...)` would avoid promotion to `Float64`. Similarly in weight initialization.

You even notice the smallest details, thanks :)

Reply:
The `Ref`s are necessary – `M` does not have a length, and you don't want to broadcast `yold`.

Reply:
I am not yet sure why one should disable mutation, but I switched to a mutating one for now.

Reply:
OK, we can think about non-mutating mean and median when it's clear that we need it. The `CurveSpace` manifold in `FunManifolds.jl` doesn't support mutation but I doubt using `mean` or `median` there is a good idea.

Reply:
And if needed, the step back to non-mutating is not that much to do.